### ISSUE TITLE
BUG: "scipy.optimize.differential_evolution" can't handle parallel computing properly 

### ISSUE THREAD

**Issue by lhg1992** [scipy/scipy#18236]
### Describe your issue.

Issue:
after setting "workers" in  "scipy.optimize.differential_evolution", it can't work any more. The issue should be the same as the following topic on stackoverflow.
https://stackoverflow.com/questions/8804830/python-multiprocessing-picklingerror-cant-pickle-type-function.

Improvement suggestion:
instead of using "from multiprocessing import Pool" which requires "if __name__ == '__main__':", it's just better to use "from concurrent.futures import ThreadPoolExecutor" for parallel computing.
Following is how I handle parallel computing at the same calling level where "differential_evolution" fails.
```python
from concurrent.futures import ThreadPoolExecutor
...
obj= lambda x: ObjL(np.block([ [PK[0], PK[1], x[0]] ]),
                            np.array([ PC ]), np.array([ VK ]), 
                            np.array([ VC ]), np.array([ [r] ]))
fun_con= lambda x: nonlconNW( ....)
...
if parallel:
    con_t=[None]*N
    obj_t=[None]*N
    pool = ThreadPoolExecutor(max_workers=workers)
    for i in range(N):
        obj_t[i] = pool.submit(obj, ([x0_list[i]]))
        con_t[i] = pool.submit(fun_con, ([x0_list[i]]))
    for i in range(N):
        obj_b[i] = obj_t[i].result()
        con_b[i] = con_t[i].result()

...
```

### Reproducing Code Example

```python
obj= lambda x: ObjL(np.block([ [PK[0], PK[1], x[0]] ]),
                            np.array([ PC ]), np.array([ VK ]), 
                            np.array([ VC ]), np.array([ [r] ]))

res = opt.differential_evolution(obj, bounds, constraints=cons, workers=4)
```

### Error message

```shell
Can't pickle <function <lambda> at 0x000002B2BE820EE0>: attribute lookup <lambda> on __main__ failed

_pickle.PicklingError: Can't pickle <function <lambda> at 0x000002B2BE820EE0>: attribute lookup <lambda> on __main__ failed
```

### SciPy/NumPy/Python version and system information

```shell
scipy ver: 1.10.0
```

**Comment by andyfaff** [comment]
Without a MWE that exercise the issue we can't look into the issue in any detail.

In any case, `ThreadPoolExecutor` is unlikely to aid any calculation speedup. This is because TPE uses threads, those threads can only run if the GIL is not being held, which means splitting calculation between threads has no use.

In comparison creating another `Pool` (`ProcessPoolExecutor`?) uses different processes which allows those parallel calculations to take place. This requires pickleability of the code concerned, and the possible use of `if __name__ == '__main__'`. (the reasons for the latter are [here](https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming)).

The documentation for differential_evolution states that you can "supply a map-like callable". So while the default is to use `multiprocessing.Pool.map` you can actually use any map-like function. e.g. `concurrent.futures.Executor.map` may be usable (but I bet you don't get any improvement in speed).

**Comment by andyfaff** [comment]
> I guess a map-like callable is something like `map(obj_fun, x_list)`, right

You would use `workers=map`. This would then get called as `map(obj_fun, x_arr)`. There is an example of how to parallelise with two processes in the examples section of the docstring. That example will use `multiprocessing.Pool` under the hood. The constraint evaluation is not parallelised, but if a constraint is infeasible then `obj_fun` is not evaluated for that particular entry.

Using `multiprocessing.Pool` has overhead, which may be too much if the objective function is expensive. If that's the case then you can try using the `vectorized` keyword.

With respect to `ThreadPoolExecutor` you will only see speedup if your objective function releases the GIL. For example:

```python
def ackley(x):
    arg1 = -0.2 * np.sqrt(0.5 * (x[0] ** 2 + x[1] ** 2))
    arg2 = 0.5 * (np.cos(2. * np.pi * x[0]) + np.cos(2. * np.pi * x[1]))
    return -20. * np.exp(arg1) - np.exp(arg2) + 20. + np.e
bounds = [(-5, 5), (-5, 5)]
```

```python
>>> %timeit result = differential_evolution(ackley, bounds, seed=1, updating='deferred')
60.1 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> from concurrent.futures import ThreadPoolExecutor as TPE
>>> pool = TPE(8)
>>> %timeit result = differential_evolution(ackley, bounds, seed=1, workers=pool.map)
83.7 ms ± 131 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

If a task is CPU bound and it does not release the GIL then each thread (from TPE) will have to wait on other threads to release the GIL before continuing. In contrast CPU bound calculations benefit from `ProcessPoolExecutor` or `multiprocessing.Pool` because calculation doesn't have to wait for the GIL to be released.

If you have a concrete MWE that shows a `ThreadPoolExecutor` speeding code evaluation up then I'd be happy to see it.

**Comment by lhg1992** [comment]
Here is a simple example showing `ThreadPoolExecutor` boosting speed.
You can clearly see the efficiency is boost linearly by the number of workers.

'''python
from concurrent.futures import ThreadPoolExecutor
import time

def action(x):
    print(f'{x**2}\n')
    time.sleep(1) # 1second
    return x**2

N=10
# %%
# parallel
tic=time.time()
future=[None]*N
res=[None]*N
pool = ThreadPoolExecutor(max_workers=5)
for i in range(N):
    future[i] = pool.submit(action, i)
for i in range(N):
    res[i] = future[i].result()
toc=time.time()
print(f"time cost for parallel: {toc-tic}")

# serial
tic=time.time()
for i in range(N):
    action(i)
toc=time.time()
print(f"time cost for serial: {toc-tic}")

'''

**Comment by andyfaff** [comment]
That example is not CPU bound, and `time.sleep` releases the GIL, i.e. it doesn't represent a real life minimization problem.

**Comment by lhg1992** [comment]
It also works for my complex engineering problem, where I evaluate the objective function and constraints. The time cost is also boosted:

'''python

if brute:
    fun_con= lambda x: nonlconNW(np.block([ [PK[0], PK[1], x[0]] ]), # shape: (n,3)
                                np.array([ PC ]), np.array([ VK ]), 
                                np.array([ VC ]), np.array([ [r] ]), 
                                x, np.array([ [PK[2]] ]), 
                                necon, lay_con, check=1)
    con_b=[None]*N
    obj_b=np.empty((N,))*np.nan
    val_ind=[False]*N
    if parallel:
        con_t=[None]*N
        obj_t=[None]*N
        pool = ThreadPoolExecutor(max_workers=workers)
        for i in range(N):
            obj_t[i] = pool.submit(obj, ([x0_list[i]]))
            con_t[i] = pool.submit(fun_con, ([x0_list[i]]))
        for i in range(N):
            obj_b[i] = obj_t[i].result()
            con_b[i] = con_t[i].result()
'''
